tidy3d/plugins/adjoint/utils/penalty.py

```diff
@@ -134,3 +136,119 @@ def penalty_fn(radius):
 
         # return the average penalty over the points
         return jnp.sum(penalty_fn(rs)) / len(rs)
+
+
+class ErosionDilationPenalty(Penalty):
+    """Computes a penalty for erosion / dilation of a parameter map not being unity.
+    Accepts a parameter array normalized between 0 and 1.
+    Uses filtering and projection methods to erode and dilate the features within this array.
+    Measures the change in the array after eroding and dilating (and also dilating and eroding).
+    Returns a penalty proportional to the magnitude of this change.
+    The amount of change under dilation and erosion is minimized if the structure has large feature
+    sizes and large radius of curvature relative to the length scale.
+
+    Note
+    ----
+    For more details, refer to chapter 4 of Hammond, A., "High-Efficiency Topology Optimization
+    for Very Large-Scale Integrated-Photonics Inverse Design" (2022).
+    """
+
+    length_scale: pd.NonNegativeFloat = pd.Field(
+        ...,
+        title="Length Scale",
+        description="Length scale of erosion and dilation."
+        "Corresponds to ``radius`` in the :class:`ConicFilter` used for filtering."
+        "The parameter array is dilated and eroded by half of this value with each operation."
+        "Roughly corresponds to the desired minimum feature size and radius of curvature",
+        units=MICROMETER,
+    )
+
+    pixel_size: pd.PositiveFloat = pd.Field(
+        ...,
+        title="Pixel Size",
+        description="Size of each pixel in the array (must be the same along all dimensions)."
+        "Corresponds to ``design_region_dl`` in the :class:`ConicFilter` used for filtering.",
+        units=MICROMETER,
+    )
+
+    beta: pd.PositiveFloat = pd.Field(
+        1.0,
+        title="Projection Beta",
+        description="Strength of the ``tanh`` projection."
+        "Corresponds to ``beta`` in the :class:`BinaryProjector`."
+        "Higher values correspond to stronger discretization.",
+    )
+
+    penalty_strength = pd.Field(
+        1.0,
+        title="Penalty Strength",
+        description="Optional value to scale the returned penalty by.",
+    )
+
+    eta0: pd.PositiveFloat = pd.Field(
+        0.5,
+        title="Projection Midpoint",
+        description="Value between 0 and 1 that sets the projection midpoint. In other words, "
+        "for values of ``eta0``, the projected values are halfway between minimum and maximum. "
+        "Corresponds to ``eta`` in the :class:`BinaryProjector`.",
+    )
+
+    delta_eta: pd.PositiveFloat = pd.Field(
+        0.01,
+        title="Delta Eta Cutoff",
+        description="The binarization threshold for erosion and dilation operations "
+        "The thresholds are 0 + `delta_eta` on the low end and `1 - delta_eta` on the high end. "
+        "This value must be tuned precisely to balance binarization with differentiability. "
+        "We strongly suggest leaving to the default value.",
```

Comment:
"leaving to" -> "using"

If it must be tuned precisely, do we know if this default is good for a broad range of settings?

Reply:
I haven't tested a lot of values, so I don't know. Maybe I'll remove this "tuned precisely" comment and instead just say that "this value works well".

Will do a fabrication constraint notebook where I'll go into more details on each of these penalties so perhaps I'll discuss it there a bit more. As we talked about before, if this value is too high (eg. 0.5) then the erosion/dilation effect is reduced or eliminated. But if it's too low (eg. 0.0) then it won't handle intermediate values the way we want.

Reply:
Changed the last two sentences to "The default value balances binarization with differentiability so we strongly suggest using it unless there is a good reason to set it differently."